### ISSUE TITLE
Read the scheduler contract shape-agnostically so the V1 model-runner pin can be dropped

### ISSUE THREAD

**Issue by viktorpusTT** [tenstorrent/vllm-tt-plugin#50]
## Summary

Upstream vLLM's V2 model runner is now the default for most models, and selecting it changes the **scheduler output contract**, not just the runner internals. The plugin reads the V1 shape of that contract and would silently produce wrong output under V2, so `TTPlatform` currently pins `VLLM_USE_V2_MODEL_RUNNER=0` (`src/vllm_tt_plugin/platform.py:492`).

That pin is a stopgap. It blocks the plugin from ever using a V2-only upstream feature, and it means every rebase has to re-verify that the V1 code paths still exist. This issue tracks making the plugin read `SchedulerOutput` correctly under **both** contract shapes, with no branch on the flag, so the pin can be deleted.

This is **not** about replicating V2's implementation. V2's data plane is Triton kernels plus CUDA UVA host-memory views, neither of which exists on TT, and none of it is involved here. This is purely about how the runner mutates its persistent batch from a scheduler diff.

Companion analysis: `docs/MODEL_RUNNER_V1_VS_V2.md`, sections 6, 7, and 9.

## Background: what the contract change is

`SchedulerOutput` splits requests into two lists. `scheduled_new_reqs` carries everything needed to build a request's state from nothing. `scheduled_cached_reqs` is an incremental diff for requests the runner already tracks.

Under V1, a request preempted and later resumed stays in `scheduled_cached_reqs`, flagged by `resumed_req_ids`, with its full token history supplied separately in `all_token_ids`.

Under V2 the scheduler instead folds resumed requests into `scheduled_new_reqs` and puts the full history in `NewRequestData.prefill_token_ids`:

```python
# vllm/v1/core/sched/scheduler.py (upstream main :1194-1204)
if self.use_v2_model_runner:
    scheduled_new_reqs.extend(scheduled_resumed_reqs)
    scheduled_resumed_reqs.clear()
    new_reqs_data = [
        NewRequestData.from_request(
            req,
            req_to_new_blocks[req.request_id].get_block_ids(),
            req._all_token_ids,          # full history, V2 only
        )
        for req in scheduled_new_reqs
    ]
```

Consequences: `CachedRequestData.resumed_req_ids` is permanently `set()`, `CachedRequestData.all_token_ids` is permanently `{}`, and `Scheduler.prev_step_scheduled_req_ids` is not maintained. Everything else in `SchedulerOutput` is flag-independent.

The flag is a computed property, `VllmConfig.use_v2_model_runner` (`vllm/config/vllm.py:615`), with no setter and no platform hook. Its default resolution ends in `return is_default_v2_architecture or not model_config.is_moe`, so every non-MoE generate model gets V2 unless a feature gate refuses.

## The failure today

Assume a request with a 100-token prompt that generated 200 tokens, then got preempted, then resumed, under V2:

1. The resume arrives in `scheduled_new_reqs` with `prompt_token_ids` = 100 prompt tokens and `prefill_token_ids` = all 300 tokens.
2. `_update_states` (`model_runner.py:649`), or the lane twin (`input_batch.py:889`), overwrites `self.requests[req_id]` with `build_cached_request_state(new_req_data)`.
3. `build_cached_request_state` (`input_batch.py:52-83`) hardcodes `output_token_ids=[]` (`:80`) and never reads `prefill_token_ids`. The 200 generated tokens are lost.
4. `InputBatch.add_request` (`input_batch.py:286`) writes the prompt into `token_ids_cpu[row, :100]`, copies the now-empty `output_token_ids` after it, and sets `num_tokens[row] = 100`.
5. `prompt_lens = num_computed_tokens_cpu[rows] + num_scheduled_tokens[req]` (`model_runner.py:1078-1082`) reaches 300, while `num_tokens[row]` is 100.
6. `input_tokens = token_ids_cpu[rows, :max_prefill_tokens]` (`model_runner.py:1084-1086`) reads **zeros** for positions 100..299. The device prefills 200 zero tokens into KV. `token_ids_cpu` is `max_model_len` wide, so there is no bounds error.
7. `intermediate_prefill_mask = prompt_lens < num_tokens[rows]` is `False`, so the step is treated as a final chunk and a token is sampled against corrupt KV.
8. The request continues as though its own 200 tokens never existed. Silent wrong output, no exception.

## The design principle

Adopt upstream V2's rule as the plugin's own rule, unconditionally:

> **A request appearing in `scheduled_new_reqs` means: discard any state currently held for this `req_id`, then rebuild it entirely from that record.**

Upstream's V2 runner does exactly this and never tries to detect resumes:

```python
# vllm/v1/worker/gpu/model_runner.py (upstream main :955-995)
for new_req_data in scheduler_output.scheduled_new_reqs:
    req_id = new_req_data.req_id
    self._remove_request(req_id)                 # unconditional
    ...
    self.req_states.add_request(
        req_id=req_id,
        prompt_len=len(new_req_data.prompt_token_ids),
        all_token_ids=new_req_data.prefill_token_ids,
        num_computed_tokens=new_req_data.num_computed_tokens,
    )
    self.block_tables.append_block_ids(req_index, new_req_data.block_ids, overwrite=True)
```

Two properties make this the right target rather than a compromise:

- It is **correct under V1**, because V1 only ever puts genuinely-new ids in `scheduled_new_reqs`, so the discard is a no-op there. No dual-path branch to maintain.
- It **removes the need for resume detection**, which is not soundly possible under V2 anyway (see below).

### Why not just detect resumes

`NewRequestData.num_computed_tokens > 0` is not a resume discriminator, in either direction:

- **False positives.** The prefix-cache lookup (`scheduler.py:811`) is guarded only by `if request.num_computed_tokens == 0:` and runs for every waiting request. A genuinely new request gets `num_computed_tokens > 0` on any prefix hit (`scheduler.py:881-883`). A request promoted out of `WAITING_FOR_REMOTE_KVS` is another source.
- **False negatives.** A resume whose blocks were all evicted comes back with `num_computed_tokens == 0`.

`len(prefill_token_ids) > len(prompt_token_ids)` is closer but fails for a request preempted during prefill, which has no output tokens yet.

`SchedulerOutput.preempted_req_ids` is sound and is populated under both flags, but it is only needed for the lane-row release below, not for state rebuilding.

## What upstream already guarantees

Worth stating explicitly because it removes the hardest hypothetical. A resume cannot arrive while a decode for that request is still in flight:

- `_preempt_request` records `request.num_stale_output_tokens = request.num_in_flight_tokens` (`scheduler.py:1369`) and zeroes `num_output_placeholders` (`:1370`).
- The waiting loop **refuses to resume** a request whose stale output is still in flight and deliverable (`scheduler.py:779-789`): "Deliverable stale output still in flight: resuming now could resample a position that output later delivers. It drains within the pipeline depth."
- `update_from_output` drains the stale share in lockstep (`:1802-1807`); the output is discarded only when `drop_stale_output` is set (`:1822-1823`).

So by the time a resume appears in `scheduled_new_reqs`, every in-flight token has already been delivered to the engine and is therefore already inside `prefill_token_ids`. The plugin does not need to preserve its own in-flight token across a resume.

This also means the identity guard at `model_runner.py:2254` (`if request_states is not None and req_state is not request_states[req_idx]: continue`) skipping a token after the state object is rebuilt is **correct**, not data loss: the rebuilt state already contains that token. That deserves a test, not a code change.

### W1. Seed the token history from `prefill_token_ids`

In `build_cached_request_state` (`input_batch.py:52`), derive the generated-token history instead of hardcoding `[]`:

- Read `getattr(new_req_data, "prefill_token_ids", None)`, so the plugin still imports against an upstream old enough to lack the field.
- If not `None`, `output_token_ids = list(prefill_token_ids[len(prompt_token_ids):])`, else `[]`.
- **Copy, do not alias.** Upstream passes `req._all_token_ids` by reference (`scheduler.py:1201`), unlike the V1 `all_token_ids` path which copies (`:1512`). TT runs a `UniProcExecutor`, so engine core and worker share a process and an aliased list would keep growing under the runner.
- Assert `len(prefill_token_ids) >= len(prompt_token_ids)`.

No other change is needed for the data path. `InputBatch.add_request` (`input_batch.py:308-317`) already copies `output_token_ids` after the prompt and sets `num_tokens = request.num_tokens`, and everything downstream keys off `num_tokens` / `num_prompt_tokens`.

**Done when:** a resumed request's prefill replays prompt plus generated tokens, `token_ids_cpu[row]` contains no zero padding below `num_tokens[row]`, and `intermediate_prefill_mask` is `True` for every chunk except the last.

### W2. Make the new-request path destructive

Before building state for any `req_id` in `scheduled_new_reqs`, unconditionally drop whatever the plugin holds for it: `self.requests.pop(req_id, None)`, the persistent-batch row (`input_batch.remove_request` plus the `condense` its docstring requires, or the lane row release), and `self._req_state_slot.pop(req_id, None)`.

Upstream's comment names a second case beyond resume: streaming input update, where a request legitimately reappears as new with an updated prompt.

Order this **after** the existing `finished_req_ids` and `preempted_req_ids` handling in `_update_states`, matching upstream's `finish_requests()`-then-`add_requests()` order, because `reset_prefix_cache(reset_running_requests=True)` can put the same id in `preempted_req_ids` and `scheduled_new_reqs` in one output (`scheduler.py:2529-2531`).

**Done when:** submitting the same `req_id` as new twice in a row leaves exactly one row, one state entry, and one state slot, and the logits-processor batch update shows a matching `removed` for every `added`.

### W3. Fix the lane resume path

`TTLaneInputBatch.apply_step_plan` (`input_batch.py:898-909`) takes its `remove_request` branch only for ids in `resumed_req_ids`. Under V2 it never fires, so a resume that kept its lane row re-registers that row in `batch_update_builder.added` while it is still live, giving logits processors a second `added` with no matching `removed`. `add_request_to_row` only raises when the row holds a *different* id (`:782-784`), so this is silent. W2 applied to the lane path fixes it.

Separately, `TTLaneCoordinator._build_step_plan` (`lane_scheduler.py:478-480`) releases a resumed request's lane row from `resumed_req_ids`. The `preempted_req_ids` release two lines above (`:475`) covers almost all of it; the residual hole is a same-step preempt-and-resume, since upstream runs the running loop before the waiting loop. Derive that case from `preempted_req_ids` intersected with the ids in `scheduled_new_reqs`.

**Done when:** the lane preempt-and-resume test passes with `resumed_req_ids` forced empty, and no row is double-added.

### W4. Do not branch on the flag

After W1 to W3 the read path is flag-agnostic. If a branch ever becomes unavoidable, read `vllm_config.use_v2_model_runner` (the config property), never the env var: auto-selection turns V2 on with the env unset, and upstream tests monkey-patch `Scheduler.use_v2_model_runner` directly.

### W5. Delete the pin, last

Remove `_pin_v1_model_runner` (`platform.py:492`) and its two call sites (`:859`, `:893`), and invert or delete `tests/test_v1_model_runner_pin.py`. Only after the tests below are green in both modes.

Note that `tests/test_v1_model_runner_pin.py` stays green through every change in W1 to W4, so it must not be mistaken for coverage of any of them.

## Things that do NOT need changing

Established by reading each site, so nobody re-derives it:

- **The lane merge** (`lane_scheduler.py:111-223`) is contract-clean. `scheduled_new_reqs` is extended verbatim so `prefill_token_ids` rides through untouched; unions of empty sets are empty. The plugin never constructs `NewRequestData`, only reads it.
- **`model_runner.py:1020` and `async_decode.py:215`** compute `is_prompt` as `len(scheduled_new_reqs) > 0 or bool(resumed_req_ids) or ...`, so the first term catches V2 resumes.
- **`model_runner.py:1016` and `:1035`** subtract `resumed_req_ids` from `cached_reqs.req_ids`. Under V2 a resume is not in `cached_reqs` on its resume step, so the subtraction is vacuous. Harmless.
- **`model_runner.py:658` and `input_batch.py:898`** `resumed_from_preemption` branches become unreachable for resumes, but `apply_cached_req_state_update`'s replace-vs-append logic (`input_batch.py:86-105`) is still needed for the V1 path and must not be deleted.
- **State-slot release** (`_release_dead_state_slots`, `model_runner.py:811-839`) already keys off `finished_req_ids` and `preempted_req_ids`, both populated under both flags. `tests/test_state_slots.py` is already contract-agnostic and is the model the other tests should follow.
- **`prev_step_scheduled_req_ids`** is never read by the plugin.
- **`TTScheduler`** overrides `schedule`, not `_update_after_schedule`, so it still inherits `AsyncScheduler`'s V2-only `next_decode_eligible_step` assignment (`async_scheduler.py:45-49`). Nothing to do, but do not start overriding `_update_after_schedule` without replicating it.
- **`max_concurrent_batches`** is 2 under both flags at `pp_size == 1`, and TT asserts `pipeline_parallel_size == 1`, so the V1-vs-V2 formula difference is currently unreachable.
- **`worker.py`, `structured_output.py`, `model_input.py`** carry no scheduler-shaped field.

## Test work

The current suite would let most of this through silently.

- **`tests/test_lane_input_batch.py:168` `_new_req`** is a duck-typed `SimpleNamespace`, not a real `NewRequestData`. It is the choke point: the moment `build_cached_request_state` reads `prefill_token_ids`, this raises `AttributeError` and takes `:214`, `:239`, and `:323` with it. Add the field there **first**, so W1 fails loudly rather than silently.
- **`tests/test_lane_input_batch.py:323`** (`test_apply_step_plan_preempted_request_frees_its_row_and_keeps_its_state`) is the one real preempt-and-resume test and it drives the resume through `CachedRequestData(resumed_req_ids={"a"})`. It needs a V2 twin driving the same scenario through `scheduled_new_reqs` with `prefill_token_ids` set and `resumed_req_ids` empty, asserting identical final state.
- **Parametrize the resume tests over both contract shapes.** Upstream does exactly this (`tests/v1/core/test_scheduler.py:3449-3470` on main).
- **Add the discrimination test that does not exist:** a genuinely new request with a prefix-cache hit (`num_computed_tokens > 0`, `prefill_token_ids == prompt_token_ids`) must not be treated as a resume, and a resume with `num_computed_tokens == 0` must be handled correctly. `tests/conftest.py:84` hard-codes `enable_prefix_caching=False`, so this needs a fixture change.
- **Add the identity-guard test:** pending async decode, then preempt, then resume, asserting the final `output_token_ids` contains the in-flight token exactly once.
- `tests/test_lane_scheduler.py:312` and `:331` put bare strings in `scheduled_new_reqs`, so the merge tests are structurally blind to `NewRequestData` field changes. Either use real records or document the limitation.
- Untested machinery in the blast radius that should gain coverage as part of this: `reset_batch` and `_decode_layout_changed_since_last_decode` have no semantic test anywhere (the only occurrences in `tests/` are inert markers in `tests/test_sample_time_grammar.py:234`); `steady_decode_scheduler_invariants_met` is untested; `_prepare_model_inputs`' `prompt_lens` trichotomy is tested only for the continuation case (`tests/test_lane_input_batch.py:280`).

## Suggested order

1. Add `prefill_token_ids` to the test fakes, so W1 fails loudly.
2. W1.
3. W2, then W3.
4. The test work above, green in both modes.
5. W4, then W5.

## Versions this analysis is based on

Re-verify before acting if significant time has passed. Upstream moves fast: 325 commits landed under `vllm/v1/worker/gpu/` between 2026-02 and 2026-08.

| What | Version |
|---|---|
| `vllm-tt-plugin` | `a070eb3`, branch `viktorpus/migrate-vllm-0251` |
| Plugin's pinned upstream target | vLLM `v0.25.1` = `752a3a504485790a2e8491cacbb35c137339ad34` (2026-07-12) |
| Upstream read for the forward-looking parts | vLLM `main` = `70afdedc1081d28c3eaae53bece8292298484c86` (2026-08-17) |
| Analysis date | 2026-08-18 |

Upstream line numbers cited above are from `main` unless noted. `NewRequestData` and `CachedRequestData` are byte-identical between `v0.25.1` and `main`, and the V2 branch in `Scheduler.schedule` plus `_make_cached_request_data` are identical too, so the contract facts hold for both. The V1-only fields are labelled `MRV1-only` upstream but carry no deprecation notice, and removing `resumed_req_ids` would break every in-tree KV connector, so a both-flags-correct consumer should stay correct indefinitely.

## Related

Two pre-existing bugs were found while analysing this code path. Both are present under the current V1 pin, both are independent of this issue, and both live in code this work touches, so fix them in the same pass:

- #52 `_is_still_prefilling` bounds the replay cursor by `num_prompt_tokens` instead of `num_tokens`, misclassifying a multi-chunk resume replay as a decode step (gemma4 chunked prefill).
- #51 The decode state-slot remap is dropped on the host-sampling path while the ownership map advances.

A third candidate was investigated and refuted: `steady_decode_scheduler_invariants_met` (`async_decode.py:212-215`) omits the `has_chunked_continuation` term that `_prepare_model_inputs` has (`model_runner.py:1017`), which looks like it should let a prefill forward run with outstanding async decode readbacks. It cannot. `_decode_layout_changed_since_last_decode` is set by every persistent-batch add, remove, and condense (`model_runner.py:695`) and is cleared only in the decode branch (`:1100`, `input_batch.py:1176`), so it stays set across prefill steps and the drain at `build_model_input:1339` fires; `reset_batch` also forces `overlap_ok=False` (`async_decode.py:279`). Adding the missing term would be a clarity edit, not a fix.
